lib/foreman/command_runner.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+require 'open3'
+
+module Foreman
+  class CommandRunner
+    attr_accessor :command, :stdin, :stdout, :stderr, :status
+
+    def initialize(command_and_args, stdin)
+      self.stdin = stdin
+      self.command = command_and_args
+      # security checks: ensure no shell, absolute path, permissions and executable
+      raise(::Foreman::Exception.new(N_("Command must be array: %s"), command)) unless command.kind_of?(Array)
+      raise(::Foreman::Exception.new(N_("At least one command must be specified"))) if command.empty?
+      raise(::Foreman::Exception.new(N_("Absolute command path must be specified: %s"), bin)) unless Pathname.new(bin).absolute?
+      raise(::Foreman::Exception.new(N_("Command %s not found or unreadable"), bin)) unless File.exist?(bin)
+      raise(::Foreman::Exception.new(N_("Command %s not executable"), bin)) unless File.executable?(bin)
```

Comment:
Very strict set of checks, we make sure shell is not spawned as we want to spawn the `ct` process directly to be able to write a good SELinux rules and also to prevent shell-injection.